tools/misc/cubemanip

```diff
@@ -1,280 +1,365 @@
-
-from numpy import oldnumeric as num
-from numpy import linalg as linal
-import math
-import sys
-import re
+#--------------------------------------------------------------------------------------------------#
+#  DFTB+: general package for performing fast atomistic simulations                                #
+#  Copyright (C) 2006 - 2023  DFTB+ developers group                                               #
+#                                                                                                  #
+#  See the LICENSE file for terms of usage and distribution.                                       #
+#--------------------------------------------------------------------------------------------------#
+
+"""Script to evaluate expressions on cube files"""
+
 import argparse
+from dataclasses import dataclass, field
+import re
+import sys
+import numpy as np
+import numpy.linalg as la
 
 SCRIPT_NAME = "cubemanip"
 
-# Tolerance for positions to be treated as equal
-tolerance = 1.0e-5
+# tolerance for positions to be treated as equal
+TOLERANCE = 1.0e-5
 
 # Pattern for the argument parsing
-patVar = re.compile(r"\((\$(?P<number>\d+))\)")
+VAR_PATTERN = re.compile(r"\((\$(?P<number>\d+))\)")
 
 # Help text
-helpTxt = """
-%(name)s <expr> [ <cubeFile1> [ <cubeFile2> [ ... ]]]
+USAGE = f"""cubemanip [-h] [-d] expr cubefile [cubefile ...]
+
+Combines the passed cube files according to an expression expression to a new
+cube file. The cube files provided as arguments after the expression must be
+referenced in the expression as ($n), with n being the index of the appropriate
+file. The expression can be any arbitrary expression which is valid in python
+using the functions found in the numpy and numpy.linalg packages and results in
+a float or a float array. Unless the '-d' option is passed, the result must have
+the same shape, as the data arrays in the cube files. The references in the
+expression are substituted by the data arrays of the appropriate cube files. The
+resulting cube file will have the same header as the first input file, while its
+data array is created by evaluating the expression. The number of cube files
+passed as arguments must be higher than the maximal reference index in the
+expression.
+
+Apart of the data arrays, following variables and functions can be used in the
+expression:
+
+  origin:  origin of the grid in the first input file as (3,) shaped array
+
+  gridvecs: grid vectors in the first input file as (3, 3) shaped array
+
+  ngrids: nr. of grid units along the grid vectors as a (3,) shaped array
+
+  posx(), posy(), posz(): x-, y- and z- coordinates of each grid point
+
+Examples:
+
+  Weighted sum of three cube files:
+  {SCRIPT_NAME} '0.25 * ($1) + 0.25 * ($2) + 0.5 * ($3)' f1.cube f2.cube f3.cube
+
+  Integral of the values in a cube file:
+  {SCRIPT_NAME} -d 'sum(($1))' file1.cube
+
+  Squared norm of the data in a cube file:
+  {SCRIPT_NAME} -d 'sum(($1)**2) * det(gridvecs)' file1.cube
```

Comment:
It looks like this syntax might be able to handle circular polarisation (in addition to the linear case you've already demonstrated). Is that correct?